design/host-discovery-controller/host-discovery.md

```diff
@@ -0,0 +1,325 @@
+2. Describe how the cluster can, and cannot, manage a host without BMC
+   credentials.
+
+### Non-Goals
+
+1. Describe a way to set BMC credentials automatically for discovered
+   hosts.
+2. Attempt to make it possible to fully manage a host without BMC
+   credentials.
+
+## Proposal
+
+### User Stories
+
+#### Adding a host and automatically provisioning
+
+*As a cluster admin, I want to add a node to a cluster without
+pre-configuring the BMC and have the cluster automatically provision
+the host as a node.*
+
+Provisioning the hosts to make them nodes can take quite a while. It
+would be nice to be able to start that process for a bunch of hosts,
+then configure their BMC credentials later to enable full management
+of the hosts. There are also situations where the BMC credentials
+might never be added (proof-of-concept or test clusters that are never
+going to be “production ready”).
+
+By combining Ironic’s discovery feature, Ironic’s fast-track
+provisioning feature, MachineSet scaling, and some new work in metal3
+we should be able to make it possible for a user to:
+
+1. Attach a host to the provisioning network
+2. Power the host on
+3. Have a `BareMetalHost` created with hardware inventory details
+4. Optionally classify the host (using the
+   `hardware-classification-controller`)
+5. Automatically scale the appropriate MachineSet based on labels on
+   the host
+6. Automatically provision the host as a node in the cluster
+
+#### Adding a host and not automatically provisioning
+
+*As a cluster admin, I want to add a node to a cluster without
+pre-configuring the BMC and NOT have the cluster automatically
+provision the host as a node.*
+
+This use case is different from the previous case because the
+`MachineSet` would not scale automatically and so the host would not
+be provisioned. This gives cluster admins an opportunity to review the
+host before provisioning, to ensure that it should be a member of the
+cluster and is not misconfigured (PXE booting from the wrong
+environment) or a malicious actor.
+
+#### Replacing a faulty node
+
+*As a data center staff member, I want to replace a faulty node in a
+cluster quickly.*
+
+The idea here is to be able to take a “broken” host out of service and
+replace it with a standby host without having access to the kubernetes
+API or UI at all. If we assume the broken host is broken enough that
+it is completely offline, it should be possible to remove it from the
+rack and replace it with another physical host, which is then
+provisioned automatically.
+
+## Design Details
+
+When Ironic is configured to support its existing discovery feature,
+unknown hosts that boot the Ironic agent will have their hardware
+details registered with `ironic-inspector` and have a node created in
+the Ironic database. Typically this happens via the PXE boot
+configuration, but in the future it could happen in other ways such as
+having a bootable ISO image with the agent on it. How the agent ends
```

Comment:
It can right now, we just need to prepare an ISO with the right parameters (ironic URL, inspector URL, inspection parameters).